### DISCUSSION TITLE
Is the CompilationDatabase config key an effective replacement for --compile-commands-dir?

### DISCUSSION THREAD

**Discussion by HighCommander4** [clangd/clangd#907] [General]
From the time I started using clangd, I have been using it with the `--compile-commands-dir` command line option to point it at my project's `compile_commands.json`. (I specify it per-project, via `"clangd.arguments"` in the workspace settings of my vscode client.)

I find this option necessary particularly for the case where as I'm browsing code I end up opening a file (typically a header) outside my project directory, i.e. the situation discussed in [this FAQ entry](https://clangd.llvm.org/faq#how-do-i-fix-errors-i-get-when-opening-headers-outside-of-my-project-directory).

In clangd 12, the [CompilationDatabase](https://clangd.llvm.org/config.html#compilationdatabase) config key was added. I have a recollection of hearing this described as a "replacement" or "preferred alternative" for `--compile-commands-dir`. (And I've reflected this preference when writing the mentioned FAQ entry.)

However, during a recent Discord discussion I realized that if `CompilationDatabase` is specified in your project's `.clangd` file, it's also **scoped** to the project's directory, so it will not take effect for headers outside the project directory.

While leads me to ask... is `CompilationDatabase` in fact an effective replacement to `--compile-commands-dir`? It seems to me like it's not, particularly for the use case of headers outside the project directory.

Am I missing something, or should `--compile-commands-dir` be the preferred way of pointing clangd to the CDB, after all?

**Comment by sam-mccall** [comment]
Seems like today that's the best thing for the workflow you describe, where you work on a single project in an editor instance but it includes files outside its FS scope, and projects are preconfigured manually.

> Am I missing something, or should --compile-commands-dir be the preferred way of pointing clangd to the CDB, after all?

I wouldn't go this far though... if you work on multiple projects at the same time this method doesn't work at all.
This idea of using a clangd instance as project boundaries seems useful but also like it has sharp edges *if* your editor doesn't have a strong concept of workspace = window = clangd instance.

I can see a couple of ways to unify the idea with the config system (it seems CDB is only the most pressing among many knobs we'd want):
 - if we're in "strong workspace mode", clangd could discover `.clangd` config files from the workspace root on startup, and apply them to all files. (Rather than having this discovery only be per-file). Is this mode opt-in, opt-out, based on heuristics? not sure.
 - we could allow the editor's workspace config to inject process-wide clangd config, e.g. `--extra-config=CompileCommands: { CompilationDatabase: /path/to/cdb }` or `--extra-config-file=/path/to/workspace.clangd`. This feels like lots of layers of indirection and quoting, and is basically a more general/obfuscated --compile-commands-dir.

> However, during a recent Discord discussion I realized that if CompilationDatabase is specified in your project's .clangd file, it's also scoped to the project's directory, so it will not take effect for headers outside the project directory.

Worth noting this is *somewhat* mitigated by the TUScheduler::HeaderIncluderCache:
 - if you've had project/bar.cc open this session (and it has a compile command from a CDB)
 - and you open <some/header.h> which is transitively included by bar.cc
 - and we don't find a trustworthy command for header.h
 - then we'll transfer the bar.cc flags for parsing header.h
This is similar to how headers are typically treated in compile_commands.json, but works even if the header is not from the same project. It's not perfect but it doesn't require manual configuration and still works when multiple CDBs are in play.

**Comment by HighCommander4** [comment]
The notion of a "strong workspace mode" sounds interesting.

Perhaps we could establish a convention where clients that want to use strong workspace mode populate the `rootPath` field of the LSP `initialize` request (or `rootUri` or `workspaceFolders` or whatever the flavour du jour is) with the workspace directory, while clients that don't leave it null / empty?

**Reply by cpsauer** [reply]
FWIW, we have same use case--and take--as you, @HighCommander4

**Reply by Hebe970217** [reply]
When I tried to use vscode to open the folder for the project, I found that many functions of Clangd are confused with the boundary of the project. 
As Sam said, Clangd is a one-to-one association with the project, which will help limit the boundary and reduce many mistakes. 
This is a great practice

**Reply by Yurii-huang** [reply]
I checked the practice of CCLS, which binds the project to the whole backend.

In my implementation, workspace and language are used as the key information of Clangd.

To achieve this, I extended the initialization request of LSP.



```
param{

language:"C/C++";

workspaceFolders{

...

}

}
```

**Reply by HighCommander4** [reply]
Somewhat related, but I only realized recently that in vscode, you can leverage our plugin's [pre-defined variables support](https://github.com/clangd/vscode-clangd/blob/37ba1e9d54501e0c694b4e9acf3f5f3cac078687/src/config.ts#L41-L48) to put the following into your vscode **user** settings (shared by all projects):

```json
"clangd.arguments": [
  "--compile-commands-dir=${workspaceFolder}"
]
```

and this allows every project to run with the project's own workspace folder as the `--compile-commands-dir` without needing any per-project configuration.

This doesn't accomplish _eveything_ we'd want a "strong workspace mode" to do (it won't apply `<workspaceFolder>/.clangd` to files outside of the workspace folder), but it gets you part of the way there.
